src/server/pfs/s3/server.go

```diff
@@ -22,74 +54,136 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		// matches /repo/branch/path/to/file.txt
 		repo := parts[1]
 		branch := parts[2]
 		file := parts[3]
-		h.serveFile(w, r, repo, branch, file)
+		if r.Method == http.MethodGet {
+			h.serveFile(w, r, repo, branch, file)
+		} else if r.Method == http.MethodPut {
+			h.putFile(w, r, repo, branch, file)
+		} else {
+			writeMethodNotAllowed(w)
+		}
 	} else {
-		http.Error(w, "not found", 404)
+		writeNotFound(w)
 	}
 }
 
 func (h handler) serveRoot(w http.ResponseWriter, r *http.Request, repo string) {
 	if err := r.ParseForm(); err != nil {
-		http.Error(w, fmt.Sprintf("%v", err), 400)
+		writeBadRequest(w, err)
 		return
 	}
 
 	if _, err := h.pc.InspectRepo(repo); err != nil {
-		code := 500
-		if strings.Contains(err.Error(), "not found") {
-			code = 404
-		}
-		http.Error(w, fmt.Sprintf("%v", err), code)
+		writeMaybeNotFound(w, err)
 		return
 	}
-	
+
 	if _, ok := r.Form["location"]; ok {
 		w.Header().Set("Content-Type", "application/xml")
 		w.Write([]byte(locationResponse))
 	} else {
-		w.Write([]byte("OK"))
+		writeOK(w)
 	}
 }
 
 func (h handler) serveFile(w http.ResponseWriter, r *http.Request, repo, branch, file string) {
 	fileInfo, err := h.pc.InspectFile(repo, branch, file)
 	if err != nil {
-		code := 500
-		if strings.Contains(err.Error(), "not found") {
-			// captures both missing branches and missing files
-			code = 404
-		}
-		http.Error(w, fmt.Sprintf("%v", err), code)
+		writeMaybeNotFound(w, err)
 		return
 	}
 
 	timestamp, err := types.TimestampFromProto(fileInfo.Committed)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("%v", err), 500)
+		writeServerError(w, err)
 		return
 	}
 
 	reader, err := h.pc.GetFileReadSeeker(repo, branch, file)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("%v", err), 500)
+		writeServerError(w, err)
 		return
 	}
 
 	http.ServeContent(w, r, "", timestamp, reader)
 }
 
+func (h handler) putFile(w http.ResponseWriter, r *http.Request, repo, branch, file string) {
+	expectedHash := r.Header.Get("Content-MD5")
+
+	if expectedHash != "" {
+		expectedHashBytes, err := base64.StdEncoding.DecodeString(expectedHash)
+
+		if err != nil {
+			writeBadRequest(w, fmt.Errorf("could not decode `Content-MD5`, as it is not base64-encoded"))
+			return
+		}
+
+		h.putFileVerifying(w, r, repo, branch, file, expectedHashBytes)
+		return
+	}
+
+	h.putFileUnverified(w, r, repo, branch, file)
+}
+
+func (h handler) putFileVerifying(w http.ResponseWriter, r *http.Request, repo, branch, file string, expectedHash []byte) {
+	hasher := md5.New()
+	reader := io.TeeReader(r.Body, hasher)
+
+	_, err := h.pc.PutFileOverwrite(repo, branch, file, reader, 0)
+	if err != nil {
+		// the error may be because the repo or branch does not exist -
+		// double-check that by inspecting the branch, so we can serve a 404
+		// instead
+		_, inspectError := h.pc.InspectBranch(repo, branch)
+		writeMaybeNotFound(w, inspectError)
+		return
+	}
+
+	actualHash := hasher.Sum(nil)
+
+	if !reflect.DeepEqual(expectedHash, actualHash) {
```

Comment:
I think `reflect.DeepEqual` is always going to return false here because `expectedHash` is a `string` and `actualHash` is a `[]byte`. To properly compare these you'll want to convert `expectedHash` into a `[]byte` and pass them to `bytes.Equal`.

Reply:
They're both byte arrays, and tests would be failing if it didn't work since minio sets `Content-MD5`. `bytes.Equal` sounds like it would be better than `reflect.DeepEqual` though